### ISSUE TITLE
When Class is converted to function - naming issues, Typescript.

### ISSUE THREAD

**Issue by RyannGalea** [evanw/esbuild#1456]
Hi all,

I'm having a particular issue when compiling that renames my classes that a package called 'type-graphql' uses, This is a package that uses decorators to decorate a class & then it generates a Graphql schema once the server initializes, which I finally got to compile after two days of messing around. Great, but the issue is it is renaming the classes by appending a 2 on them. see code here, and compiled code below.

```
@Service()
@Resolver() 
export class AddBookingMutation {

  @Authorized()
  @UseMiddleware(HasRole([UserRolesEnum.ADMIN, UserRolesEnum.BOOKING_ADD], RoleMessagesEnum.BOOKING_ADD))
  @UseMiddleware(AutoReference(EntitiesEnum.BOOKING))
  @Mutation(() => Booking)
  @CatchAll('Add Booking Save Error')
  async addBooking(
    @Arg('input', () => BookingInput) input: BookingInput
  ): Promise<Booking | Error> {

    const { _id } = await BookingModel.create({...input})
    return BookingModel
      .findById(_id)
      .populate('assignedTo')
      .populate('job')
      .exec()

  }

}

```

Renamed to "AddBookingMutation2"
```
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddBookingMutation = void 0;
var auth_middleware_1 = require("../../../core/middleware/auth.middleware");
var user_roles_enum_1 = require("../../../core/enums/user-roles.enum");
var booking_entity_1 = require("../../../entity/job/booking.entity");
var type_graphql_1 = require("type-graphql");
var auto_ref_middleware_1 = require("../../../core/middleware/auto-ref.middleware");
var entities_enum_1 = require("../../../core/enums/entities.enum");
var typedi_1 = require("typedi");
var catch_all_helper_1 = require("../../../core/helpers/catch-all.helper");
var role_messages_enum_1 = require("../../../core/enums/role-messages.enum");
var AddBookingMutation = function() {
  function AddBookingMutation2() {
  }
  AddBookingMutation2.prototype.addBooking = function(input) {
    return __awaiter(this, void 0, Promise, function() {
      var _id;
      return __generator(this, function(_a) {
        switch (_a.label) {
          case 0:
            return [4, booking_entity_1.BookingModel.create(__assign({}, input))];
          case 1:
            _id = _a.sent()._id;
            return [2, booking_entity_1.BookingModel.findById(_id).populate("assignedTo").populate("job").exec()];
        }
      });
    });
  };
  __decorate([
    type_graphql_1.Authorized(),
    type_graphql_1.UseMiddleware(auth_middleware_1.HasRole([user_roles_enum_1.UserRolesEnum.ADMIN, user_roles_enum_1.UserRolesEnum.BUSINESS_ADD], role_messages_enum_1.RoleMessagesEnum.BOOKING_ADD)),
    type_graphql_1.UseMiddleware(auto_ref_middleware_1.AutoReference(entities_enum_1.EntitiesEnum.BOOKING)),
    type_graphql_1.Mutation(function() {
      return booking_entity_1.Booking;
    }),
    catch_all_helper_1.CatchAll("Add Booking Save Error"),
    __param(0, type_graphql_1.Arg("input", function() {
      return booking_entity_1.BookingInput;
    }))
  ], AddBookingMutation2.prototype, "addBooking", null);
  AddBookingMutation2 = __decorate([
    typedi_1.Service(),
    type_graphql_1.Resolver()
  ], AddBookingMutation2);
  return AddBookingMutation2;
}();
exports.AddBookingMutation = AddBookingMutation;

```

The problem is the clientside currently rely on the schema being accurate with the original resolver names, and I don't think it's a great idea to append 2 on all my clientside code.

I feel as if this should be a relatively easy fix to the compilation settings but thus far haven't been able to find any good info on it, especially with esbuild. It seems there might be a plugin for webpack but I haven't investigated that too much as I have no experience with it.

Is there any way I can stop esbuild doing this renaming?

My tsconfig looks like:
```
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es2018",
    "moduleResolution": "node",
    "rootDir": "src",
    "baseUrl": "src",
    "outDir": "dist",
    "sourceMap": false,
    "declaration": false,
    "paths": {	
      "@core/*": ["./core/*"],	
      "@entity/*": ["./entity/*"],	
      "@api/*": ["./api/*"]	
    },
    "lib": [  
      "dom",
      "es2018",
      "esnext.asynciterable"
    ],
    "skipLibCheck": true,
    "importHelpers": true,
    "esModuleInterop": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "allowSyntheticDefaultImports": true
  },
  "include": [
    "src"
  ],
  "exclude": [
    "dist",
    "node_modules"
  ]
}
```

```
[watch:graphql]     GraphQLDocumentError: Unknown type "Job". Did you mean "Job2"?
[watch:graphql]         at C:/outbuild/client/projects/outbuild/src/app/data/graphql/sendgrid/sendgrid.graphql:76:25
[watch:graphql]     GraphQLDocumentError: Unknown type "Parcel". Did you mean "Parcel2" or "Price2"?
[watch:graphql]         at C:/outbuild/client/projects/outbuild/src/app/data/graphql/sendgrid/sendgrid.graphql:21:28
[watch:graphql]     GraphQLDocumentError: Unknown type "Property". Did you mean "Property2"?
[watch:graphql]         at C:/outbuild/client/projects/outbuild/src/app/data/graphql/sendgrid/sendgrid.graphql:68:30
[watch:graphql]     GraphQLDocumentError: Unknown type "Supplier". Did you mean "Supplier2"?
[watch:graphql]         at C:/outbuild/client/projects/outbuild/src/app/data/graphql/sendgrid/sendgrid.graphql:37:30
[watch:graphql]     GraphQLDocumentError: Unknown type "Times". Did you mean "Time2", "Times2", "File2", or "Item2"?
[watch:graphql]         at C:/outbuild/client/projects/outbuild/src/app/data/graphql/sendgrid/sendgrid.graphql:1:27
```

Any advice or ideas are welcome, a little stuck on this.

**Comment by hyrious** [comment]
It seems like the result is not directly produced by esbuild. Instead, it could be produced by:
```
source code
-> tsc --target=es5 --module=commonjs // `class` becomes `function`
-> esbuild // the `AddBookingMutation` name in the function renamed to …2
```

**Comment by RyannGalea** [comment]
Not sure where to go on this,

**Comment by gozmanyoni** [comment]
Did you try using https://esbuild.github.io/api/#keep-names ?

**Comment by RyannGalea** [comment]
> Did you try using https://esbuild.github.io/api/#keep-names ?

I did try that, I believe 'keepNames' only works with minification. Like @hyrious mentioned it may happen during the 'tsc' step. Although I have transpiled it standalone with just 'tsc' and it doesn't rename the class.

**Comment by RyannGalea** [comment]
This is the output of the function if I just use 'tsc'. I guess it doesn't do the transformation into a function.

```
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddBookingMutation = void 0;
const tslib_1 = require("tslib");
const auth_middleware_1 = require("@core/middleware/auth.middleware");
const user_roles_enum_1 = require("@core/enums/user-roles.enum");
const booking_entity_1 = require("@entity/job/booking.entity");
const type_graphql_1 = require("type-graphql");
const auto_ref_middleware_1 = require("@core/middleware/auto-ref.middleware");
const entities_enum_1 = require("@core/enums/entities.enum");
const typedi_1 = require("typedi");
const catch_all_helper_1 = require("@core/helpers/catch-all.helper");
const role_messages_enum_1 = require("@core/enums/role-messages.enum");
let AddBookingMutation = class AddBookingMutation {
    async addBooking(input) {
        const { _id } = await booking_entity_1.BookingModel.create({ ...input });
        return booking_entity_1.BookingModel
            .findById(_id)
            .populate('assignedTo')
            .populate('job')
            .exec();
    }
};
tslib_1.__decorate([
    type_graphql_1.Authorized(),
    type_graphql_1.UseMiddleware(auth_middleware_1.HasRole([user_roles_enum_1.UserRolesEnum.ADMIN, user_roles_enum_1.UserRolesEnum.BOOKING_ADD], role_messages_enum_1.RoleMessagesEnum.BOOKING_ADD)),
    type_graphql_1.UseMiddleware(auto_ref_middleware_1.AutoReference(entities_enum_1.EntitiesEnum.BOOKING)),
    type_graphql_1.Mutation(() => booking_entity_1.Booking),
    catch_all_helper_1.CatchAll('Add Booking Save Error'),
    tslib_1.__param(0, type_graphql_1.Arg('input', () => booking_entity_1.BookingInput)),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [booking_entity_1.BookingInput]),
    tslib_1.__metadata("design:returntype", Promise)
], AddBookingMutation.prototype, "addBooking", null);
AddBookingMutation = tslib_1.__decorate([
    typedi_1.Service(),
    type_graphql_1.Resolver()
], AddBookingMutation);
exports.AddBookingMutation = AddBookingMutation;
```

**Comment by hyrious** [comment]
let me clarify what happens here,
to name a class or function, we have similar ways:
```js
var A = class {}
var A = function() {}
// all produces A.name === 'A'
```
however, typescript's es5 will convert a class to:
```js
var A = /** @class */ (function(){
  function A() {} // ←
  return A
})()
```
then esbuild convert the inner function's name to A2 to prevent name conflict, which causes different name.

I don't think esbuild should keep inner functions' names. In this situation, I'll recommend you to emit an 'esnext' code then lower it to es5 using babel or similar tools.

**Comment by RyannGalea** [comment]
@hyrious Thank you for taking the time to write back. It gave me the direction I needed to get this resolved,

I set up my build to compile esnext - 'tsc', then it runs 'tsc-alias' to resolve some paths I have set up. I then use Babel to convert it down to es5 & minify, I also need to add a setting to minify 'keepFnName' which allows the code to be minified without ruining the schema naming. I also use a plugin called  "@babel/plugin-transform-runtime" to help with some async-await stuff.

This is currently compiling for me now.

I think I will continue to use esbuild for development only.
